### PR TITLE
feat: setRange support list type

### PR DESCRIPTION
**Pull request by lpxxn**
# PR Details

## Description

SetRange支持了[]string
修改了一点代码中的问题，比如 _XXType 写在一起了

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: datavalidation.go [hunk 1]
```diff
@@ -23,12 +23,12 @@ type DataValidationType int
...
+	_        DataValidationType = iota
+	typeNone                    // inline use
 	DataValidationTypeCustom
 	DataValidationTypeDate
 	DataValidationTypeDecimal
-	typeList // inline use
+	DataValidationTypeList // inline use
 	DataValidationTypeTextLength
 	DataValidationTypeTime
 	// DataValidationTypeWhole Integer
```
Comment (xuri) [inline datavalidation.go:34]: Unused constant `DataValidationTypeWhole` can be removed.

File: datavalidation.go [hunk 2]
```diff
@@ -23,12 +23,12 @@ type DataValidationType int
...
 const (
-	_DataValidationType = iota
-	typeNone            // inline use
+	_        DataValidationType = iota
+	typeNone                    // inline use
 	DataValidationTypeCustom
 	DataValidationTypeDate
 	DataValidationTypeDecimal
-	typeList // inline use
+	DataValidationTypeList // inline use
```
Comment (xuri) [inline datavalidation.go:31]: The comments `// inline use` can be removed

File: datavalidation.go [hunk 3]
```diff
@@ -119,18 +124,26 @@ func (dv *DataValidation) SetInput(title, msg string) {
...
-	dv.Formula1 = fmt.Sprintf(`<formula1>"%s"</formula1>`, formulaEscaper.Replace(formula))
-	dv.Type = convDataValidationType(typeList)
-	return nil
+	return fmt.Sprintf(`<%[2]s>"%[1]s"</%[2]s>`, formulaEscaper.Replace(formula), formulaName), nil
 }
 
 // SetRange provides function to set data validation range in drop list, only
-// accepts int, float64, or string data type formula argument.
-func (dv *DataValidation) SetRange(f1, f2 interface{}, t DataValidationType, o DataValidationOperator) error {
+// accepts int, float64, string or []string data type formula argument.
```
Comment (xuri) [inline datavalidation.go:145]: There already exists a function `SetDropList` that can be used for creating drop list data validation, so I think we needn't add the same support for the `SetRange` function.

File: datavalidation.go [hunk 4]
```diff
@@ -155,6 +174,11 @@ func (dv *DataValidation) SetRange(f1, f2 interface{}, t DataValidationType, o D
 		formula2 = fmt.Sprintf("<formula2>%.17g</formula2>", v)
 	case string:
 		formula2 = fmt.Sprintf("<formula2>%s</formula2>", v)
+	case []string:
```
Comment (xuri) [inline datavalidation.go:177]: When do we need to set `formula2` as a list? This case has no test cases covered.

**Review by xuri — CHANGES_REQUESTED** [review]
Thanks for your pull request. I've left some comments.

File: datavalidation.go [hunk 5]
```diff
@@ -142,9 +155,15 @@ func (dv *DataValidation) SetRange(f1, f2 interface{}, t DataValidationType, o D
 		formula1 = fmt.Sprintf("<formula1>%.17g</formula1>", v)
 	case string:
 		formula1 = fmt.Sprintf("<formula1>%s</formula1>", v)
+	case []string:
+		formula1, err = dv.extractListFormula(v, formula1Name)
+		if err != nil {
+			return err
```
Comment (xuri) [inline datavalidation.go:161]: This error case hasn't been covered.

File: datavalidation.go [hunk 6]
```diff
@@ -142,9 +155,15 @@ func (dv *DataValidation) SetRange(f1, f2 interface{}, t DataValidationType, o D
 		formula1 = fmt.Sprintf("<formula1>%.17g</formula1>", v)
 	case string:
 		formula1 = fmt.Sprintf("<formula1>%s</formula1>", v)
+	case []string:
+		formula1, err = dv.extractListFormula(v, formula1Name)
```
Comment (xuri) [inline datavalidation.go:159]: This can be simplify to:

```go
if formula1, err = dv.extractListFormula(v, formula1Name); err != nil {
	return err
}
```

**Review by xuri — CHANGES_REQUESTED** [review]

File: datavalidation.go (hunk 3 above)
Comment (lpxxn) [reply datavalidation.go:145]: OK

**Comment by xuri** [comment]
Hi @lpxxn, would you like to reopen this pull request and resolve the suggestion as I mentions in this code review?

**Review by xuri — APPROVED** [review]
LGTM. Thanks for your contribution. I have made some changes based on your branch.

**Pull request merged**